### PR TITLE
Trigger page register when module is executed

### LINKED ISSUE #244: Importing a css file in _app.js may cause Cannot read property 'default' of undefined
```
next@6.1.1-canary.5
next-css@0.2.1-canary.1
```

I started seeing a pretty magic bug in production recently and am writing to share the results of my investigation. Here's what was showing up in the browser's console during some of the page loads:

```
TypeError: Cannot read property 'default' of undefined
    at http://localhost:3000/_next/static/bc4f7954-f656-4e9e-af24-a4404cecb3c7/pages/_app.js:1:4762
    at r (http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:9:96586)
    at e.value (http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:9:96973)
    at http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:1:1034
    at Array.forEach (<anonymous>)
    at Object.110 (http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:1:991)
    at l (http://localhost:3000/_next/static/runtime/webpack-7dba61e9b6510e660c00.js:1:520)
    at Object.105 (http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:1:156)
    at l (http://localhost:3000/_next/static/runtime/webpack-7dba61e9b6510e660c00.js:1:520)
    at Object.104 (http://localhost:3000/_next/static/runtime/main-68ba217b0c0184534eba.js:1:103)
```

This was followed by no JavaScript working on the client. However, if the exception was somehow avoided on the first page load, all subsequent Next.js pages were working well as long as they were loaded dynamically via `<Link/>`. The same exception never showed up in development.

After removing pretty much any bit from my app (😅) and testing various hypotheses, I was able to craft an MWE where the exception was reproduced pretty stably. All you need to do is to install `next-css`, create `pages/_app.js` and import a CSS file there:

```jsx
import App, { Container } from "next/app";
import React from "react";

import "../styles/index.css"; // 😱

export default class MyApp extends App {
  render() {
    const { Component, pageProps } = this.props;
    return (
      <Container>
        <Component {...pageProps} />
      </Container>
    );
  }
}
```

Interestingly, the exception was only occurring in Chrome and not Firefox and could disappear when I varied the size and the complexity of a page. However, with `pages/index.jsx` being just `export default () => <div>hello</div>;` the exception appeared every time.

Here's how adding `import "../styles/index.css";` to `_app.jsx` affects the HTML in production:

```diff
 <!DOCTYPE html>
 <html>
 
 <head>
   <meta charSet="utf-8" class="next-head" />
   <link rel="preload" href="/_next/static/HASH/pages/index.js" as="script" />
   <link rel="preload" href="/_next/static/HASH/pages/_app.js" as="script" />
   <link rel="preload" href="/_next/static/HASH/pages/_error.js" as="script" />
   <link rel="preload" href="/_next/static/runtime/webpack-HASH.js" as="script" />
   <link rel="preload" href="/_next/static/chunks/HASH.js" as="script" />
   <link rel="preload" href="/_next/static/runtime/main-HASH.js" as="script" />
+  <link rel="preload" href="/_next/static/chunks/HASH.js" as="script" />
+  <link rel="stylesheet" href="/_next/static/css/styles.HASH.chunk.css" />
 </head>
 
 <body>
   <div id="__next">
     <div>hello</div>
   </div>
   <div id="__next-error"></div>
   <script>
     __NEXT_DATA__ = { "props": { "pageProps": {} }, "page": "/", "pathname": "/", "query": {}, "buildId": "HASH" }
     module = {}
     __NEXT_LOADED_PAGES__ = []
 
     __NEXT_REGISTER_PAGE = function (route, fn) {
       __NEXT_LOADED_PAGES__.push({ route: route, fn: fn })
     }
   </script>
   <script async="" id="__NEXT_PAGE__/" src="/_next/static/HASH/pages/index.js"></script>
   <script async="" id="__NEXT_PAGE__/_app" src="/_next/static/HASH/pages/_app.js"></script>
   <script async="" id="__NEXT_PAGE__/_error" src="/_next/static/HASH/pages/_error.js"></script>
   <script src="/_next/static/runtime/webpack-HASH.js" async=""></script>
   <script src="/_next/static/chunks/HASH.js" async=""></script>
   <script src="/_next/static/runtime/main-HASH.js" async=""></script>
+  <script src="/_next/static/chunks/HASH.js" async=""></script>
 </body>
 
</html>
```

As you can see, importing a stylesheet in `_app.js` not just adds a link to a CSS file, but also produces a new JS chunk. This JS file does not contain anything really:

```js
(window.webpackJsonp=window.webpackJsonp||[]).push([[6],{220:function(n,w,o){}}]);
```

`/_next/static/HASH/pages/_app.js` changes a bit too and my suspicion is that it is trying to import a module from the extra JS chunk before that one is loaded. This race condition does not happen on any page because the size of its server-rendered content may affect the order in which Chrome loads the JS files and bootstraps the client-side app.

Getting rid of the bug was pretty simple. I just moved the CSS import into the `page` HOC in my real app (`pages/index.js` in the MWE) and the exception never showed up again.

What can we do to save others from the same trap? It took me more than half a day to detect the bug and to investigate it 😅

### LINKED ISSUE #4620: protocol+domain+port gets added to asPath after a HMR
# Bug report

If I pass down `ctx.ctx.asPath` and print it in my `<Component />` and then force a HMR, it'll add the protocol+domain+port to the beginning of `asPath`.

```
{
  "url": {
    "query": {},
    "pathname": "/",
    "asPath": "/"
  }
}

// After a bundle rebuild

{
  "url": {
    "query": {},
    "pathname": "/",
    "asPath": "http://localhost:3000/"
  }
}
```

## To Reproduce
`pages/_app.js`:

```
import App, { Container } from 'next/app';
import React from 'react';

export default class MyApp extends App {
  static async getInitialProps ({ Component, router, ctx }) {
    const { query, pathname, asPath } = ctx;
    return { url: { query, pathname, asPath } };
  }

  render () {
    const {Component, url} = this.props;
    return (<Container><Component url={url} /></Container>);
  }
}
```
`pages/index.js`:
```
export default (props) => <pre>{JSON.stringify(props, null, 2)}</pre>;
```

Force a bundle rebuild by adding some text and save.

## System information
 - Version of Next.js: 6.0.3

## Other information
As discussed on Slack, it's related to this [line](https://github.com/zeit/next.js/blob/e881c1b3867f3a843847e1cfecbe0e393d425ae1/lib/router/router.js#L103).

### LINKED ISSUE #4997: Crash on 7.0.0-canary.0, production build
# Bug report

## Describe the bug

Production build. Sometimes when reload page, app crashes.
"var r" undefined on 1 screenshot

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Reload page, or go to some page
2. See error

## Expected behavior
Page without errors

## System information
 - OS: Linux
 - Version of Next.js: 7.0.0-canary.0

## Additional context

Error does not always occur, sometimes everything works as usual.

### PR DESCRIPTION
**Pull request by timneutkens**
Solves inconsistent loading issues when one bundle is loaded faster than the other

Fixes zeit/next-plugins#244
Fixes #4997
Fixes #4620

### CODE REVIEW

**Comment by timneutkens** [comment]
@Zn4rK I finally found the issue you ran into and wrote a test for, check this commit: https://github.com/zeit/next.js/pull/5115/commits/a3f83ae2216f47449dfd24e6e3e038664674f527

It was a timing issue where the page cache got deleted right before re-rendering the page. Causing the `props` to not be available. Removing this logic as it is no longer needed there was enough 👍

**Comment by vjpr** [comment]
@timneutkens When are you planning to merge this? Is it imminent?

Just wondering because I really need this fix and would prefer not to use my fork.

**Review by HaNdTriX — APPROVED** [review]

File: build/webpack/plugins/pages-plugin.js [hunk 1]
```diff
@@ -27,16 +40,12 @@ export default class PagesPlugin {
...
-        source.add(modules)
-        source.add(`
-            return { page: comp.default }
-          })
-        `)
+        const source = new ConcatSource(
+          `__NEXT_REGISTER_PAGE('${routeName}', function() {\n`,
+          moduleSourcePostModule,
+          '\nreturn { page: module.exports.default }',
+          '});'
```
Comment (HaNdTriX) [inline build/webpack/plugins/pages-plugin.js]: We might want to remove the linebreaks here or make them consistent.

**Comment by timneutkens** [comment]
Basically today, was just waiting for Someone to review which has been done now.

**Pull request merged**